File: sdks/nodejs/intercept.js
'use strict';

const http = require('http');
const https = require('https');
const { URL } = require('url');
const { EventEmitter } = require('events');

/**
 * Baafoo Full SDK - Node.js HTTP module interception
 *
 * Replaces http.request and https.request to intercept all outbound HTTP requests.
 * Supports stub, record, record-and-stub, and passthrough modes.
 */

let _originalHttpRequest = null;
let _originalHttpsRequest = null;
let _originalHttpGet = null;
let _originalHttpsGet = null;
let _patched = false;
let _sdkInstance = null;

/**
 * 启用 HTTP 拦截：monkey-patch http.request 和 https.request
 * @param {object} client - Baafoo Client 实例
 */
function patch(client) {
  if (_patched) return;
  _patched = true;
  _sdkInstance = client;

  _originalHttpRequest = http.request;
  _originalHttpsRequest = https.request;
  _originalHttpGet = http.get;
  _originalHttpsGet = https.get;

  http.request = _patchedRequest;
  https.request = _patchedRequest;
  // http.get 内部缓存了原始 request 引用，必须单独 patch
  http.get = function(...args) {
    const req = _patchedRequest(...args);
    req.end();
    return req;
  };
  https.get = function(...args) {
    const req = _patchedRequest(...args);
    req.end();
    return req;
  };
}

/**
 * 恢复原始的 http.request 和 https.request
 */
function unpatch() {
  if (!_patched) return;
  _patched = false;
  _sdkInstance = null;

  if (_originalHttpRequest) {
    http.request = _originalHttpRequest;
  }
  if (_originalHttpsRequest) {
    https.request = _originalHttpsRequest;
  }
  if (_originalHttpGet) {
    http.get = _originalHttpGet;
  }
  if (_originalHttpsGet) {
    https.get = _originalHttpsGet;
  }
  _originalHttpRequest = null;
  _originalHttpsRequest = null;
  _originalHttpGet = null;
  _originalHttpsGet = null;
}

/**
 * 拦截后的 request 实现
 */
function _patchedRequest(...args) {
  if (!_sdkInstance) {
    return _originalHttpRequest(...args);
  }

  // 解析参数
  let options, callback;
  if (typeof args[0] === 'string' || args[0] instanceof URL) {
    const parsed = new URL(args[0]);
    options = {
      method: 'GET',
      hostname: parsed.hostname,
      port: parsed.port || (parsed.protocol === 'https:' ? 443 : 80),
      path: parsed.pathname + parsed.search,
      headers: {},
    };
    if (args[1] && typeof args[1] === 'object') {
      Object.assign(options, args[1]);
    }
    callback = typeof args[args.length - 1] === 'function' ? args[args.length - 1] : null;
  } else if (typeof args[0] === 'object') {
    options = { ...args[0] };
    callback = typeof args[args.length - 1] === 'function' ? args[args.length - 1] : null;
  } else {
    return _originalHttpRequest(...args);
  }

  const method = (options.method || 'GET').toUpperCase();
  const path = options.path || '/';
  const host = options.hostname || options.host || 'localhost';
  let port = parseInt(options.port, 10) || 80;

  const mode = _sdkInstance.mode;
  const rules = _sdkInstance.rules;

  // 在 stub/record-and-stub 模式下尝试匹配规则
  let matchedRule = null;
  if (mode === 'stub' || mode === 'record-and-stub') {
    for (const rule of rules) {
      if (!rule.enabled) continue;
      if (_matchRule(rule, method, path)) {
        matchedRule = rule;
        break;
      }
    }
  }

  if (matchedRule) {
    // 返回 mock 响应
    return _buildMockResponse(matchedRule, options, callback, method, path, host, port);
  }

  // 未匹配规则：passthrough 或 record 模式
  const startTime = Date.now();

  // 包装 callback 以录制响应
  const wrappedCallback = (res) => {
    const durationMs = Date.now() - startTime;

    if (mode === 'record' || mode === 'record-and-stub') {
      _recordResponse(_sdkInstance, method, path, host, port, res, durationMs, options);
    }

    if (callback) {
      callback(res);
    }
  };

  // 使用原始的 request
  const newArgs = [...args];
  if (newArgs.length > 0 && typeof newArgs[newArgs.length - 1] === 'function') {
    newArgs[newArgs.length - 1] = wrappedCallback;
  } else {
    newArgs.push(wrappedCallback);
  }

  return _originalHttpRequest.apply(http, newArgs);
}

/**
 * 规则匹配
 */
function _matchRule(rule, method, path) {
  for (const cond of (rule.conditions || [])) {
    if (cond.field === 'method') {
      if (!_matchOperator(cond.operator, method, cond.value)) return false;
    } else if (cond.field === 'path') {
      if (!_matchOperator(cond.operator, path, cond.value)) return false;
    }
  }
  return true;
}

/**
 * 操作符匹配
 */
function _matchOperator(operator, actual, expected) {
  switch (operator) {
    case 'equals': case '=': return actual === expected;
    case 'contains': return actual.includes(expected);
    case 'prefix': return actual.startsWith(expected);
    case 'suffix': return actual.endsWith(expected);
    case 'exists': return actual !== '';
    default: return actual === expected;
  }
}

/**
 * 构建 mock 响应并返回一个模拟的 ClientRequest
 */
function _buildMockResponse(rule, options, callback, method, path, host, port) {
  const respConfig = (rule.responses && rule.responses[0]) || {};
  const statusCode = respConfig.statusCode || 200;
  const body = respConfig.body || '';
  const headers = respConfig.headers || {};
  const delayMs = respConfig.delayMs || 0;

  // 创建一个模拟的 IncomingMessage
  const res = new http.IncomingMessage(/* socket */ null);
  res.statusCode = statusCode;
  res.headers = {};
  for (const [k, v] of Object.entries(headers)) {
    res.headers[k] = v;
    res.headers[k.toLowerCase()] = v;
  }

  // 创建一个模拟的 ClientRequest（基于 EventEmitter，不发起真实连接）
  const req = new EventEmitter();
  req.finished = false;
  req.writableEnded = false;
  req.destroyed = false;
  req.write = function(_chunk, _enc, _cb) { return true; };
  req.end = function(_chunk, _enc, _cb) {
    this.writableEnded = true;
    this.finished = true;
    return this;
  };
  req.abort = function() { this.destroyed = true; this.emit('abort'); };
  req.destroy = function(_err) {
    this.destroyed = true;
    if (_err) this.emit('error', _err);
    return this;
  };
  req.setTimeout = function(_ms, _cb) { return this; };
  req.setNoDelay = function(_n) { return this; };
  req.setSocketKeepAlive = function(_e, _d) { return this; };
  req.flushHeaders = function() {};

  // 延迟发送响应
  const sendResponse = () => {
    // 如果是 record-and-stub 模式，录制请求
    if (_sdkInstance.mode === 'record-and-stub') {
      _sdkInstance.reportRecording({
        protocol: 'http',
        host: host,
        port: port,
        method: method,
        path: path,
        responseStatusCode: statusCode,
        responseHeaders: headers,
        responseBody: body,
        responseTimeMs: delayMs,
        direction: 'request',
      });
    }

    // 先发送响应头事件，再调用 callback（让调用方注册 res 的 data/end 监听器）
    req.emit('response', res);
    if (callback) {
      callback(res);
    }

    // 异步发送响应体（确保调用方已注册 data/end 监听器）
    const emitBody = () => {
      if (body) {
        const bodyBuf = Buffer.from(body);
        res.emit('data', bodyBuf);
      }
      res.emit('end');
    };
    // 使用 process.nextTick 确保监听器已注册
    process.nextTick(emitBody);
  };

  if (delayMs > 0) {
    setTimeout(sendResponse, delayMs);
  } else {
    process.nextTick(sendResponse);
  }

  return req;
}

/**
 * 录制响应
 */
function _recordResponse(sdk, method, path, host, port, res, durationMs, reqOptions) {
  const statusCode = res.statusCode || 200;
  const respHeaders = {};
  for (const [k, v] of Object.entries(res.headers || {})) {
    respHeaders[k] = v;
  }

  // 收集响应体
  const bodyChunks = [];
  res.on('data', (chunk) => {
    bodyChunks.push(chunk);
  });
  res.on('end', () => {
    const responseBody = Buffer.concat(bodyChunks).toString('utf-8');

    sdk.reportRecording({
      protocol: 'http',
      host: host,
      port: port,
      method: method,
      path: path,
      requestHeaders: reqOptions.headers || {},
      responseStatusCode: statusCode,
      responseHeaders: respHeaders,
      responseBody: responseBody,
      responseTimeMs: durationMs,
      direction: 'request',
    });
  });
}

module.exports = { patch, unpatch };

File: sdks/python/baafoo/intercept.py
"""Baafoo Full SDK - Python HTTP interception via monkey-patching.

Replaces urllib.request.urlopen and http.client.HTTPConnection to intercept
all outbound HTTP requests. Supports stub, record, record-and-stub, and
passthrough modes.
"""

import io
import time
import urllib.request
from urllib.parse import urlparse

from .models import RecordingEntry


# 保存原始函数的引用
_original_urlopen = urllib.request.urlopen
_original_http_connection_init = None
_patched = False
_sdk_instance = None


def patch(client):
    """启用 HTTP 拦截：monkey-patch urllib.request.urlopen

    Args:
        client: baafoo.Client 实例
    """
    global _patched, _sdk_instance
    if _patched:
        return
    _patched = True
    _sdk_instance = client
    urllib.request.urlopen = _patched_urlopen


def unpatch():
    """恢复原始的 urllib.request.urlopen"""
    global _patched, _sdk_instance
    if not _patched:
        return
    _patched = False
    _sdk_instance = None
    urllib.request.urlopen = _original_urlopen


def _patched_urlopen(url, data=None, timeout=None, **kwargs):
    """拦截后的 urlopen 实现"""
    if _sdk_instance is None:
        return _original_urlopen(url, data, timeout, **kwargs)

    # 解析 URL
    if isinstance(url, str):
        parsed = urlparse(url)
        method = "GET" if data is None else "POST"
        path = parsed.path or "/"
        host = parsed.hostname or ""
        port = parsed.port or (443 if parsed.scheme == "https" else 80)
    elif hasattr(url, "full_url"):
        # Request 对象
        parsed = urlparse(url.full_url)
        method = url.get_method()
        path = parsed.path or "/"
        host = parsed.hostname or ""
        port = parsed.port or (443 if parsed.scheme == "https" else 80)
        data = url.data if data is None else data
    else:
        return _original_urlopen(url, data, timeout, **kwargs)

    mode = _sdk_instance.mode
    rules = _sdk_instance.rules

    # 在 stub/record-and-stub 模式下尝试匹配规则
    matched_rule = None
    if mode in ("stub", "record-and-stub"):
        for rule in rules:
            if not rule.enabled:
                continue
            if _match_rule(rule, method, path):
                matched_rule = rule
                break

    if matched_rule:
        # 返回 mock 响应
        resp = _build_mock_response(matched_rule, method, path)
        if mode == "record-and-stub":
            _record_request(_sdk_instance, method, path, host, port,
                            data, resp, 0)
        return resp

    # 未匹配规则：passthrough 或 record 模式
    start_time = time.time()
    resp = _original_urlopen(url, data, timeout, **kwargs)
    duration_ms = int((time.time() - start_time) * 1000)

    if mode in ("record", "record-and-stub"):
        _record_request(_sdk_instance, method, path, host, port,
                        data, resp, duration_ms)

    return resp


def _match_rule(rule, method, path):
    """规则匹配"""
    for cond in rule.conditions:
        if cond.field == "method":
            if not _match_operator(cond.operator, method, cond.value):
                return False
        elif cond.field == "path":
            if not _match_operator(cond.operator, path, cond.value):
                return False
    return True


def _match_operator(operator, actual, expected):
    """操作符匹配"""
    if operator in ("equals", "="):
        return actual == expected
    elif operator == "contains":
        return expected in actual
    elif operator == "prefix":
        return actual.startswith(expected)
    elif operator == "suffix":
        return actual.endswith(expected)
    elif operator == "exists":
        return actual != ""
    else:
        return actual == expected


def _build_mock_response(rule, method, path):
    """构建 mock 响应"""
    if not rule.responses:
        return _MockResponse(200, {}, b"")

    resp_config = rule.responses[0]
    status_code = resp_config.status_code or 200
    body = resp_config.body or ""
    body_bytes = body.encode("utf-8") if isinstance(body, str) else body

    # 延迟
    if resp_config.delay_ms > 0:
        time.sleep(resp_config.delay_ms / 1000.0)

    return _MockResponse(status_code, resp_config.headers, body_bytes)


def _record_request(sdk, method, path, host, port, req_data, resp, duration_ms):
    """录制请求/响应"""
    req_body = ""
    if req_data:
        req_body = req_data.decode("utf-8") if isinstance(req_data, bytes) else str(req_data)

    resp_body = ""
    resp_status = 200
    resp_headers = {}

    if resp is not None:
        if hasattr(resp, "status"):
            resp_status = resp.status
        elif hasattr(resp, "code"):
            resp_status = resp.code

        if hasattr(resp, "headers"):
            resp_headers = dict(resp.headers)
        elif hasattr(resp, "info"):
            try:
                resp_headers = dict(resp.info())
            except Exception:
                pass

        # 读取响应体
        if hasattr(resp, "read"):
            try:
                body_bytes = resp.read()
                resp_body = body_bytes.decode("utf-8", errors="replace")
                # 恢复可读状态
                resp.read = lambda: body_bytes
            except Exception:
                pass

    entry = RecordingEntry(
        protocol="http",
        host=host,
        port=port,
        method=method,
        path=path,
        request_body=req_body,
        response_status_code=resp_status,
        response_headers=resp_headers,
        response_body=resp_body,
        response_time_ms=duration_ms,
        direction="request",
    )
    sdk.report_recording(entry)


class _MockResponse:
    """模拟 HTTP 响应对象，兼容 urllib 的 addinfourl 接口"""

    def __init__(self, status_code, headers, body_bytes):
        self.status = status_code
        self.code = status_code
        self._body = body_bytes
        self._read = False
        self.headers = {}
        for k, v in (headers or {}).items():
            self.headers[k] = v

    def read(self, amt=None):
        if self._read:
            return b""
        self._read = True
        if amt is None:
            return self._body
        return self._body[:amt]

    def info(self):
        return _DictHeaders(self.headers)

    def getcode(self):
        return self.status

    def geturl(self):
        return ""

    def __enter__(self):
        return self

    def __exit__(self, *args):
        pass


class _DictHeaders:
    """简单的 headers 字典包装"""

    def __init__(self, headers):
        self._headers = headers or {}

    def __getitem__(self, key):
        return self._headers.get(key, "")

    def get(self, key, default=None):
        return self._headers.get(key, default)

    def items(self):
        return self._headers.items()

    def keys(self):
        return self._headers.keys()
